File: client/common/packet.go
package common

import (
	"bytes"
	"encoding/binary"
	"errors"
)

type Packet struct {
	Nombre     string
	Apellido   string
	Nacimiento string
	Documento  uint32
	Numero     uint32
}

func NewPacket(nombre, apellido, nacimiento string, documento, numero uint32) (*Packet, error) {
	if len(nombre) > 30 || len(apellido) > 30 {
		return nil, errors.New("no debe superar los 30 caracteres")
	}
	if len(nacimiento) != 10 {
		return nil, errors.New("formato de fecha incorrecto")
	}
	return &Packet{nombre, apellido, nacimiento, documento, numero}, nil
}

func (packet *Packet) Serialize() ([]byte, error) {
	var buf bytes.Buffer

	nombreBytes := []byte(packet.Nombre)
	apellidoBytes := []byte(packet.Apellido)
	nacimientoBytes := []byte(packet.Nacimiento)

	log.Infof("nombreBytes: %v", nombreBytes)
	log.Infof("apellidoBytes: %v", apellidoBytes)
	log.Infof("nacimientoBytes: %v", nacimientoBytes)
	buf.WriteByte(byte(len(nombreBytes)))
	buf.Write(nombreBytes)

	buf.WriteByte(byte(len(apellidoBytes)))
	buf.Write(apellidoBytes)

	buf.Write(nacimientoBytes)

	if err := binary.Write(&buf, binary.BigEndian, packet.Documento); err != nil {
		return nil, err
	}
	if err := binary.Write(&buf, binary.BigEndian, packet.Numero); err != nil {
		return nil, err
	}

	payload := buf.Bytes()
	payloadLength := make([]byte, 2)
	binary.BigEndian.PutUint16(payloadLength, uint16(len(payload)))

	return append(payloadLength, payload...), nil
}

func Deserialize(data_bytes []byte) (*Packet, error) {
	if len(data_bytes) < 2 {
		return nil, errors.New("formato invalido de paquete")
	}

	data := data_bytes[2:]

	// Lectura de nombre
	nombreLength := int(data[0])
	nombre := string(data[1 : 1+nombreLength])

	// Lectura de apellido
	apellidoLengthField := 1 + nombreLength
	apellidoLength := int(data[apellidoLengthField])
	apellidoStart := apellidoLengthField + 1
	apellido := string(data[apellidoStart : apellidoStart+apellidoLength])

	// Lectura de nacimiento
	nacimientoStart := apellidoStart + apellidoLength
	nacimiento := string(data[nacimientoStart : nacimientoStart+10])

	// Lectura de documento
	documentoStart := nacimientoStart + 10
	documento := binary.BigEndian.Uint32(data[documentoStart : documentoStart+4])

	// Lectura de numero
	numeroStart := documentoStart + 4
	numero := binary.BigEndian.Uint32(data[numeroStart : numeroStart+4])

	return &Packet{nombre, apellido, nacimiento, documento, numero}, nil
}

File: server/common/packet.py
class Packet:
    def __init__(self, nombre, apellido, nacimiento, documento, numero):
        if len(nombre) > 30 or len(apellido) > 30:
            raise ValueError("No debe superar los 30 caracteres.")
        if len(nacimiento) != 10:
            raise ValueError("Formato de fecha incorrecto.")
        
        self.nombre = nombre
        self.apellido = apellido
        self.nacimiento = nacimiento
        self.documento = documento
        self.numero = numero

    def serialize(self):
        # Nombre y apellido a bytes
        nombre_bytes = self.nombre.encode('utf-8')
        apellido_bytes = self.apellido.encode('utf-8')

        # Longitud de los nombres en bytes
        nombre_length_bytes = len(nombre_bytes).to_bytes(1, byteorder='big', signed=False)
        apellido_length_bytes = len(apellido_bytes).to_bytes(1, byteorder='big', signed=False)

        # Fecha de nacimiento a bytes, es de 10 caracteres siempre y todos en rango ASCII
        nacimiento_bytes = self.nacimiento.encode('utf-8')

        # Documento y numero a bytes
        documento_bytes = self.documento.to_bytes(4, byteorder='big', signed=False)
        numero_bytes = self.numero.to_bytes(4, byteorder='big', signed=False)

        payload = nombre_length_bytes + nombre_bytes + apellido_length_bytes + apellido_bytes + nacimiento_bytes + documento_bytes + numero_bytes
        payload_length = len(payload).to_bytes(2, byteorder='big', signed=False)
        
        return payload_length + payload

    @classmethod
    def deserialize(cls, byte_data):
        byte_data = byte_data[2:]
        
        # Lectura de nombre
        nombre_length_bytes = int.from_bytes(byte_data[0:1], byteorder='big', signed=False)
        nombre_bytes = byte_data[1:1 + nombre_length_bytes]
        nombre = nombre_bytes.decode('utf-8')

        # Lectura de apellido
        apellido_length_field = 1 + nombre_length_bytes
        apellido_length_bytes = int.from_bytes(byte_data[apellido_length_field:apellido_length_field + 1], byteorder='big', signed=False)
        
        apellido_start = apellido_length_field + 1
        apellido_bytes = byte_data[apellido_start:apellido_start + apellido_length_bytes]
        apellido = apellido_bytes.decode('utf-8')
        
        # Lectura de nacimiento
        nacimiento_start = apellido_start + apellido_length_bytes
        nacimiento_bytes = byte_data[nacimiento_start:nacimiento_start + 10]
        nacimiento = nacimiento_bytes.decode('utf-8')

        # Lectura de documento
        documento_start = nacimiento_start + 10
        documento_bytes = byte_data[documento_start:documento_start + 4]
        documento = int.from_bytes(documento_bytes, byteorder='big', signed=False)

        # Lectura de numero
        numero_start = documento_start + 4
        numero_bytes = byte_data[numero_start:numero_start + 4]
        numero = int.from_bytes(numero_bytes, byteorder='big', signed=False)
               
        return cls(nombre, apellido, nacimiento, documento, numero)